src/core/Object3D.js

```diff
@@ -411,27 +406,58 @@ class Object3D extends EventDispatcher {
 
 	attach( object ) {
 
-		// adds object as a child of this, while maintaining the object's world transform
+		if ( arguments.length > 1 ) {
+
+			for ( let i = 0; i < arguments.length; i ++ ) {
 
-		// Note: This method does not support scene graphs having non-uniformly-scaled nodes(s)
+				this.attach( arguments[ i ] );
 
-		this.updateWorldMatrix( true, false );
+			}
 
-		_m1.copy( this.matrixWorld ).invert();
+			return this;
 
-		if ( object.parent !== null ) {
+		}
 
-			object.parent.updateWorldMatrix( true, false );
+		if ( object === this ) {
 
-			_m1.multiply( object.parent.matrixWorld );
+			console.error( 'THREE.Object3D.attach: object can\'t be attached as a child of itself.', object );
+			return this;
 
 		}
 
-		object.applyMatrix4( _m1 );
+		if ( object && object.isObject3D ) {
+
+			// adds object as a child of this, while maintaining the object's world transform
+
+			// Note: This method does not support scene graphs having non-uniformly-scaled node(s)
+
+			this.updateWorldMatrix( true, false );
 
-		this.add( object );
+			_m1.copy( this.matrixWorld ).invert();
 
-		object.updateWorldMatrix( false, true );
+			if ( object.parent !== null ) {
+
+				object.parent.updateWorldMatrix( true, false );
+
+				_m1.multiply( object.parent.matrixWorld );
+
+			}
+
+			object.applyMatrix4( _m1 );
+
+			object.removeFromParent();
+			object.parent = this;
+			this.children.push( object );
+
+			object.updateWorldMatrix( false, true );
+
+			object.dispatchEvent( _addedEvent );
```

Comment:
Do you mind keep the usage of `add()` in this method? It should be sufficient if you just add the following code snippet at the top of the method:
```js
if ( arguments.length > 1 ) {

	for ( let i = 0; i < arguments.length; i ++ ) {

		this.attach( arguments[ i ] );

	}

	return this;

}
```

Reply:
if so, `added` will emit too early
```js
this.add( .. ) //  <<< emitted `added` already
object.updateWorldMatrix( false, true );
```
also, misleading error msg: `THREE.Object3D.add: ...` (should be `.attach`)

so, nope :)

Reply:
Ah okay, I wasn't aware of these issues. Nevertheless, it does not feel good that so much code is copy/pasted. I wonder if we can solve this differently.